File: src/sapi.cpp
#include <iostream>
#include <string>
#include "FarmAnimal.h"
#include "sapi.h"
#include "Milkproducing.h"
#include "Meatproducing.h"

using namespace std;
const int a = 10;
const int b = 4;
const int c = 5;
const int d = 6;

sapi::sapi(int posX, int posY):FarmAnimal(posX, posY), Meatproducing(posX, posY), Milkproducing(posX, posY){
    Full=FULLMAX;
    countcowMilk=0;
    this->posX = FarmAnimal::getPosX();
    this->posY = FarmAnimal::getPosY();
}

sapi::~sapi(){
    this->posX = 0;
    this->posY = 0;
    Full = 0;
    countcowMilk = 0;
    
    //cout<<"One of your cows has died"<<endl;
}

void sapi::move(Cell&_c) {
    int xa = FarmAnimal::getPosX();
    int x = rand()%(a+b-1);
    int y = rand()%(a+b-1);
    _c.setElement(posX,posY,' ');
    //cout<<" x , y "<<x<< " " <<y<<endl;
    xa+=x;
    this->posY +=y;
    while(xa>a) {
         x = rand()%(a+b-1);
        //cout<<" x "<<x<<endl;
        while (this->posY>b) {
            this->posY -=x;
        }
        this->posX -=x;
        if (xa<a) {
            xa = xa + a;
        }
    }
    FarmAnimal::setPosX(xa);
    _c.setElement(posX,posY,'A');
    cout<<"pos MOVE : "<< this->posX<<" " << this->posY<<endl;
    
    Full--;
}

void sapi::talk(){
    cout<<"Moooo mooooo"<<endl;
}
void sapi::eat(Cell& _c){
    if (_c.getRumput(this->posX,this->posY)){
        Full=FULLMAX;
        countcowMilk++;
        if (countcowMilk >MAX){
            countcowMilk = MAX;
        }
        _c.setElement(posX,posY,' ');
    }
}

string sapi::getChar() {
    return "S";
}

string sapi::getProduct(){
    return "";
}

//void setProduct(string cowMilk);
int sapi::getcountcowMilk() const{
    return countcowMilk;
}

int sapi::getFull(){
    return Full;
}

void sapi::setFull(int Full){
    this->Full=Full;
}

void sapi::Print(){
    cout<<"CM ada " <<countcowMilk<<endl;
    cout<<"Full : "<< Full<<endl;
    cout<<"pos : "<< this->posX<<" " << this->posY<<endl;
}

/*void sapi::setPosX(int x){
    posX = x;
}

void sapi::setPosY(int y){
    posY= y;
}

int sapi ::getPosX(){
    return posX;
}

int sapi ::getPosY(){
    return posY;
}
*/

File: java/sapi.java
import java.util.concurrent.ThreadLocalRandom;

// nextInt is normally exclusive of the top value,
// so add 1 to make it inclusive

public class sapi extends FarmAnimal implements LivingThings{
   char t_rumput ='#';
   char tempat ='-';
   char produk ='U';

    final int randX =1;
    final int randY =1;

    //private int Full;
    private int posX;
    private int posY;
    private int countcowMilk;

    //Renderable c;

    public sapi (int posX ,int posY,Renderable c){
       
        super(posX,posY);
       // c.setElement(posX,posY)
        //Cell c = new Cell();
        c.setElement(posX,posY,'S');
        this.countcowMilk =0;
        this.posX = posX;
        this.posY = posY;
        //this.Full = FULLMAX;


    }
    public void move (Renderable c ){
        int x = ThreadLocalRandom.current().nextInt(-randX, randY + 1);
        int y =ThreadLocalRandom.current().nextInt(-randX, randY);
        //System.out.println("x : " + x);
        //System.out.println("y : " +y);
        
        if (this.posX+x>=0 && this.posY+y>=0 && this.posY +y <15 && hidup() && c.getElement(this.posX+x,this.posY+y)== ' '){
            
            if (c.getTypeLand(this.posX+x,this.posY+y)==tempat){
                c.setElement(posX,posY,' ');
                this.posX+=x;
                this.posY +=y;
                c.setElement(this.posX,this.posY,'S');
                setFull(getFull()-1);
                
            } 
            
            //System.out.println(getFull());
            //c.setElement(this.posX,this.posY,'S');
        } else {
            if (!hidup()){
                c.setElement(this.posX,this.posY,' ');
            }
            
        }
    }
    public void talk(){
        System.out.println("MOO MOO");
    }
    public void eat(Renderable c){
        if (!hidup()){
                //setStatus(false);
                c.setElement(this.posX,this.posY,' ');
        } else {
            if(c.getRumput(this.posX,this.posY)==t_rumput){
                this.countcowMilk =MAX;
                setFull(FULLMAX);
                c.kosongRumput(this.posX,this.posY);
         } else move(c);
        
        }
       
    }
    public String getChar(){
        return "S";
    }

    public void getProduct(){
        this.countcowMilk =0;
    }
    public Integer getCountProduct(){
        return countcowMilk;
    }
    /*public Integer getFull(){
        return Full;
    }*/

    /*public void setFull(int Full){
        this.Full = Full;
    }*/
    public void Print(){
        System.out.println("sapi " + countcowMilk + " Full : "+ getFull()+"pos :  "+ posX +" "+ posY);
    }
}
